### PR TITLE
fancy lighting

### PR DESCRIPTION
**Pull request by harryob**
this was a long time coming

many thanks to:
https://github.com/tgstation/TerraGov-Marine-Corps/pull/1964 for the lighting controller (A-lexa)
https://github.com/tgstation/TerraGov-Marine-Corps/pull/4747/ and https://github.com/tgstation/TerraGov-Marine-Corps/pull/7263 for the lighting (TiviPlus)
https://github.com/tgstation/tgstation/pull/54520 for the dir lighting component
tgstation/tgstation#75018 for the out of bounds fix in lighting
https://github.com/tgstation/TerraGov-Marine-Corps/pull/6678 for the emissives (TiviPlus) (but this is still a 
- [ ] to do)

:cl: TiviPlus, A-lexa, SyncIt21, Azarak, harryob, MorrowWolf, Zonespace
add: shadows are now in the game, and light follows you.
/:cl:

### CODE REVIEW

**Comment by Segrain** [comment]
Trying to construct an APC breaks.
image
image

**Comment by harryob** [comment]
> Trying to construct an APC breaks. image image

fixed

**Comment by Time-Warper** [comment]
Hello, as you can see in these images, the flares and the guns with rail light can be seen through the walls. I suspect the problem has something to do with the this merge
Lighting bug 1
Lighting bug 2

It also happened in the last rounds i played.

**Comment by Nanu308** [comment]
Fixed cameras having lights on them making them hard to see/click and vice versa. Solved with commit [7a2b8a4]

File: code/game/area/areas.dm [hunk 1]
```diff
@@ -28,8 +28,6 @@
 	var/unique = TRUE
 
 	var/has_gravity = 1
-	var/area/master // master area used for power calcluations
-								// (original area before splitting due to sd_DAL)
 	var/list/related // the other areas of the same type as this
```
Comment (fira) [inline code/game/area/areas.dm:31]: `related` is pretty much the counterpart to `master`, believe it can be stripped away with the rest

**Review by fira — COMMENTED** [review]
Exciting!

File: code/game/atoms_movable.dm [hunk 2]
```diff
@@ -21,14 +21,29 @@
...
-/atom/movable/Destroy()
+/atom/movable/Destroy(force)
 	for(var/atom/movable/I in contents)
 		qdel(I)
 	if(pulledby)
 		pulledby.stop_pulling()
 	QDEL_NULL(launch_metadata)
 
+	if(opacity)
+		RemoveElement(/datum/element/light_blocking)
```
[the remark this answers is not in the document]
Comment (harryob) [reply code/game/atoms_movable.dm:43]: that is a very good point

File: code/modules/mob/living/living.dm [hunk 3]
```diff
@@ -16,6 +16,8 @@
 
 	attack_icon = image("icon" = 'icons/effects/attacks.dmi',"icon_state" = "", "layer" = 0)
 
+	sight |= SEE_BLACKNESS
```
Comment (kalazus) [inline code/modules/mob/living/living.dm:19]: Looks like this breaks thermo vision (blackness rendered over game objects like mobs), but requires further check.

File: code/datums/components/overlay_lighting.dm [hunk 4]
```diff
@@ -0,0 +1,306 @@
...
+		set_holder(movable_parent)
+		return
+	var/atom/inside = movable_parent.loc //Parent's loc
+	if(isnull(inside))
+		set_holder(null)
+		return
+	if(isturf(inside.loc))
+		set_holder(inside)
+		return
+	set_holder(null)
```
Comment (Drulikar) [inline code/datums/components/overlay_lighting.dm:192]: For the issue where lights pre-equipped in cryopods don't work, now the light will still attach even if we are deeper inside something's contents (cryopod).
```suggestion
	if(!movable_parent || !get_turf(movable_parent))
		set_holder(null)
		return
	// Try only going up to parent one level
	if(!isturf(movable_parent.loc) && !ismob(movable_parent))
		movable_parent = movable_parent.loc
	set_holder(movable_parent)
```
You may also consider adding more checks to either movable_parent or movable_parent.loc for any other scenarios where you don't want parent getting illuminated when inserted. E.g. `&& !istype(movable_parent.loc, /obj/structure/closet)`

**Comment by mrbadassgent** [comment]
Possible bug: When a xenomorph swallows me for transport and spits me out after I have turned the light my flashlight grip on. The light from my gun will not render even though it has turned on and is still turned on. The light would render in old lighting, so I assume it's not intentional to change this.

**Pull request merged**